src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -890,9 +890,9 @@
   {(schema/required-key "certname") schema/Str
    (schema/required-key "environment") schema/Str
    (schema/required-key "code_ast") schema/Str
-   (schema/required-key "trusted_facts") {(schema/required-key "values") {schema/Str schema/Any}}
-   (schema/required-key "facts") {(schema/required-key "values") {schema/Str schema/Any}}
-   (schema/required-key "variables") {(schema/required-key "values") {schema/Str schema/Any}}
+   (schema/required-key "trusted_facts") {(schema/required-key "values") schema/Str}
```

Comment:
I realize this will now break the old compile endpoint and we need to figure out a real solution. This was mainly just to see if this will even work.

Reply:
The serialization bug we are trying to avoid is that it appears that some complex objects are not passed to ruby properly https://github.com/puppetlabs/puppetserver/blob/89f3d86485cb7d1ac46e0428432b85acc69863ab/src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj#L89 Specifically we will see for example when a key in a hash points to a value that is a list of hashes ruby sees it as something like `"missing_update_kbs"=>#<Java::ClojureLang::PersistentVector:0x61cfd20e>,`. My quick workaround (https://github.com/puppetlabs/orchestrator/pull/1338/commits/39ff2e5b592ff742ae16d62af49580240b3800f2)  was to serialize as a JSON string in so its just a string when clojure passes it through to ruby. This allows ruby to just deserialize the JSON.

Reply:
That makes sense, I actually wonder why we haven't seen that in the endpoint that CD4PE uses. That's [this code](https://github.com/puppetlabs/puppetserver/blob/master/src/clj/puppetlabs/services/master/master_core.clj#L905
) fwiw.

It looks like in normal catalog compilation we pass the facts in as a string so they can be deserialized by Puppet [here](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/catalog/compiler.rb#L87). 

fwiw, I think you want to use [`Puppet::Util::Json`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/json.rb) here.

Reply:
TBH I think I might advocate to just make the breaking API change. Nothing else has used this, since no-one ever opened tickets or anything to fix the serialization problem. Since the endpoint is broken without this change I think breaking the API should be alright in this case.